microceph/ceph/nfs.go

```diff
@@ -0,0 +1,316 @@
+package ceph
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"net/http"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/canonical/lxd/shared/api"
+
+	"github.com/canonical/microceph/microceph/constants"
+	"github.com/canonical/microceph/microceph/database"
+	"github.com/canonical/microceph/microceph/interfaces"
+)
+
+// EnableNFS enables the NFS Ganesha service on the cluster and adds initial configuration.
+func EnableNFS(s interfaces.StateInterface, clusterID, bindAddress string, bindPort, v4MinVersion uint, monitorAddresses []string) error {
+	hostname, err := os.Hostname()
+	if err != nil {
+		return err
+	}
+
+	pathConsts := constants.GetPathConst()
+	ganeshaConfDir := filepath.Join(pathConsts.ConfPath, "ganesha")
+	err = os.MkdirAll(ganeshaConfDir, 0744)
+	if err != nil && !os.IsExist(err) {
+		return err
+	}
+
+	// Create NFS Ganesha configuration.
+	configs := map[string]any{
+		"bindAddr":      bindAddress,
+		"bindPort":      bindPort,
+		"confDir":       ganeshaConfDir,
+		"clusterID":     clusterID,
+		"minorVersions": v4MinVersion,
+	}
+
+	ganeshaConf := newGaneshaConfig(ganeshaConfDir)
+	err = ganeshaConf.WriteConfig(configs, 0644)
+	if err != nil {
+		return err
+	}
+
+	// Create NFS Ganesha Ceph configuration.
+	configs = map[string]any{
+		"confDir":  ganeshaConfDir,
+		"monitors": strings.Join(monitorAddresses, ","),
+	}
+
+	cephConf := newGaneshaCephConfig(ganeshaConfDir)
+	err = cephConf.WriteConfig(configs, 0644)
+	if err != nil {
+		return err
+	}
+
+	// Create NFS Ganesha Ceph keyring.
+	err = createNFSKeyring(ganeshaConfDir, clusterID)
+	if err != nil {
+		return err
+	}
+
+	// Create the NFS Pool if needed.
+	err = ensureNFSPool(clusterID)
+	if err != nil {
+		return err
+	}
+
+	// Add the node to the Shared Grace Management Database.
+	err = addNodeToSharedGraceMgmtDb(filepath.Join(ganeshaConfDir, "ceph.conf"), clusterID, hostname)
+	if err != nil {
+		return nil
+	}
+
+	// Start the NFS Ganesha service.
+	return startNFS()
+}
+
+// DisableNFS disables the NFS service on the cluster.
+func DisableNFS(ctx context.Context, s interfaces.StateInterface, clusterID string) error {
+	pathConsts := constants.GetPathConst()
+	ganeshaConfDir := filepath.Join(pathConsts.ConfPath, "ganesha")
+
+	// Stop the NFS Ganesha service.
+	err := stopNFS()
+	if err != nil {
+		return err
+	}
+
+	// Remove the NFS Ganesha Ceph keyring.
+	err = os.Remove(filepath.Join(ganeshaConfDir, "keyring"))
```

Comment:
I think we should also remove the NFS-Ganesha ceph client keyring record from the `ceph auth` keys.